File: src/core/virtualserialmanager.cpp
#include "virtualserialmanager.h"
#include <QFile>
#include <QDebug>
#include <QTextCodec>

#ifdef Q_OS_WIN
#include <QSerialPortInfo>
#else
#include <QRegularExpression>
#include <QProcess>
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>

static const char* VPORT_SYMLINK = "/tmp/mousart_vport";
#endif

VirtualSerialManager::VirtualSerialManager(QObject *parent)
    : QObject(parent)
{
#ifndef Q_OS_WIN
    connect(&m_socat, &QProcess::readyReadStandardOutput, this, &VirtualSerialManager::onSocatOutput);
    connect(&m_socat, &QProcess::readyReadStandardError, this, &VirtualSerialManager::onSocatOutput);
    connect(&m_socat, QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished),
            this, &VirtualSerialManager::onSocatFinished);
#endif

    m_timedSendTimer.setTimerType(Qt::PreciseTimer);
    connect(&m_timedSendTimer, &QTimer::timeout, this, &VirtualSerialManager::onTimedSendTick);

    // Stats timer
    m_statsTimer.start();
    m_statsUpdateTimer.setInterval(1000);
    connect(&m_statsUpdateTimer, &QTimer::timeout, this, [this]() {
        m_rxBytesSnapshot = m_rxBytes;
        m_txBytesSnapshot = m_txBytes;
        m_statsTimer.restart();
        emit statsChanged();
    });
    m_statsUpdateTimer.start();
}

VirtualSerialManager::~VirtualSerialManager()
{
    stopVirtualPort();
}

bool VirtualSerialManager::startVirtualPort()
{
#ifdef Q_OS_WIN
    emit errorOccurred(tr("虚拟串口功能在 Windows 上暂不可用，请使用硬件串口调试模式。"));
    return false;
#else
    if (m_socat.state() == QProcess::Running) {
        stopVirtualPort();
    }

    m_ptyA.clear();
    m_ptyB.clear();
    m_waitingForPorts = true;

    m_socat.start("socat", QStringList()
                     << "-d" << "-d"
                     << "pty,raw,echo=0"
                     << QString("pty,raw,echo=0,link=%1").arg(VPORT_SYMLINK));

    if (!m_socat.waitForStarted(3000)) {
        m_waitingForPorts = false;
        emit errorOccurred(tr("启动 socat 失败: ") + m_socat.errorString());
        return false;
    }

    return true;
#endif
}

void VirtualSerialManager::stopVirtualPort()
{
    stopTimedSend();

#ifdef Q_OS_WIN
    if (m_winPort) {
        m_winPort->close();
        delete m_winPort;
        m_winPort = nullptr;
        emit isActiveChanged();
        emit externalPortChanged();
    }
#else
    m_waitingForPorts = false;

    if (m_readNotifier) {
        m_readNotifier->setEnabled(false);
        delete m_readNotifier;
        m_readNotifier = nullptr;
    }

    if (m_fd >= 0) {
        ::close(m_fd);
        m_fd = -1;
    }

    if (m_socat.state() == QProcess::Running) {
        m_socat.terminate();
        if (!m_socat.waitForFinished(2000)) {
            m_socat.kill();
            m_socat.waitForFinished(1000);
        }
    }

    m_ptyA.clear();
    m_ptyB.clear();
    QFile::remove(VPORT_SYMLINK);
    emit isActiveChanged();
    emit externalPortChanged();
#endif
}

qint64 VirtualSerialManager::sendData(const QString &data, bool hexMode)
{
#ifdef Q_OS_WIN
    if (!m_winPort || !m_winPort->isOpen()) {
        emit errorOccurred(tr("虚拟串口未打开"));
        return -1;
    }

    QByteArray bytes;
    if (hexMode) {
        QString hex = QString(data).remove(' ').remove('\n').remove('\r');
        bytes = QByteArray::fromHex(hex.toUtf8());
    } else {
        QString text = data;
        if (m_newlineCr) text += "\r";
        if (m_newlineLf) text += "\n";
        bytes = text.toUtf8();
    }

    qint64 written = m_winPort->write(bytes);
    if (written == -1) {
        emit errorOccurred(m_winPort->errorString());
    } else {
        m_txBytes += written;
        emit statsChanged();
    }
    return written;
#else
    if (m_fd < 0) {
        emit errorOccurred(tr("虚拟串口未打开"));
        return -1;
    }

    QByteArray bytes;
    if (hexMode) {
        QString hex = QString(data).remove(' ').remove('\n').remove('\r');
        bytes = QByteArray::fromHex(hex.toUtf8());
    } else {
        QString text = data;
        if (m_newlineCr) text += "\r";
        if (m_newlineLf) text += "\n";
        bytes = text.toUtf8();
    }

    qint64 written = ::write(m_fd, bytes.constData(), bytes.size());
    if (written == -1) {
        emit errorOccurred(QString::fromLocal8Bit(strerror(errno)));
    } else {
        m_txBytes += written;
        emit statsChanged();
    }
    return written;
#endif
}

void VirtualSerialManager::startTimedSend(const QString &data, bool hexMode, int intervalMs, int count)
{
    if (!isActive()) {
        emit errorOccurred(tr("虚拟串口未打开"));
        return;
    }
    m_timedSendData = data;
    m_timedSendHexMode = hexMode;
    m_timedSendCount = count;
    m_timedSendSent = 0;
    m_timedSendTimer.setInterval(qMax(intervalMs, 1));
    m_timedSendTimer.start();
    emit timedSendActiveChanged();
}

void VirtualSerialManager::stopTimedSend()
{
    m_timedSendTimer.stop();
    m_timedSendSent = 0;
    m_timedSendCount = -1;
    emit timedSendActiveChanged();
}

void VirtualSerialManager::onTimedSendTick()
{
    if (!isActive()) {
        stopTimedSend();
        return;
    }
    qint64 result = sendData(m_timedSendData, m_timedSendHexMode);
    if (result > 0) {
        m_timedSendSent++;
        emit timedSendCompleted(m_timedSendData);
        if (m_timedSendCount > 0 && m_timedSendSent >= m_timedSendCount) {
            stopTimedSend();
        }
    }
}

// Stats
double VirtualSerialManager::rxRate() const
{
    qint64 elapsed = m_statsTimer.elapsed();
    if (elapsed < 100) return 0.0;
    return (double)(m_rxBytes - m_rxBytesSnapshot) * 1000.0 / elapsed;
}

double VirtualSerialManager::txRate() const
{
    qint64 elapsed = m_statsTimer.elapsed();
    if (elapsed < 100) return 0.0;
    return (double)(m_txBytes - m_txBytesSnapshot) * 1000.0 / elapsed;
}

void VirtualSerialManager::resetStats()
{
    m_rxBytes = 0;
    m_txBytes = 0;
    m_rxBytesSnapshot = 0;
    m_txBytesSnapshot = 0;
    m_statsTimer.restart();
    emit statsChanged();
}

// Settings
void VirtualSerialManager::setEchoEnabled(bool enabled)
{
    if (m_echoEnabled == enabled) return;
    m_echoEnabled = enabled;
    emit echoEnabledChanged();
}

void VirtualSerialManager::setNewlineCr(bool v)
{
    if (m_newlineCr == v) return;
    m_newlineCr = v;
    emit newlineChanged();
}

void VirtualSerialManager::setNewlineLf(bool v)
{
    if (m_newlineLf == v) return;
    m_newlineLf = v;
    emit newlineChanged();
}

void VirtualSerialManager::setReceiveEncoding(const QString &enc)
{
    if (m_receiveEncoding == enc) return;
    m_receiveEncoding = enc;
    emit receiveEncodingChanged();
}

QString VirtualSerialManager::decodeData(const QByteArray &data) const
{
    if (m_receiveEncoding == "UTF-8") {
        return QString::fromUtf8(data);
    } else if (m_receiveEncoding == "GBK" || m_receiveEncoding == "GB18030") {
        QTextCodec *codec = QTextCodec::codecForName(m_receiveEncoding.toUtf8());
        if (codec) return codec->toUnicode(data);
        return QString::fromLocal8Bit(data);
    } else if (m_receiveEncoding == "Latin-1") {
        return QString::fromLatin1(data);
    } else if (m_receiveEncoding == "ASCII") {
        QString result;
        for (char c : data) {
            if (c >= 0x20 && c <= 0x7E) result += QChar(c);
            else result += QString("\\x%1").arg((unsigned char)c, 2, 16, QChar('0'));
        }
        return result;
    }
    return QString::fromUtf8(data);
}

#ifndef Q_OS_WIN
void VirtualSerialManager::onSocatOutput()
{
    QByteArray output = m_socat.readAllStandardOutput() + m_socat.readAllStandardError();
    QString text = QString::fromUtf8(output);

    if (!m_waitingForPorts) return;

    QRegularExpression re(R"(PTY is (\S+))");
    QRegularExpressionMatchIterator it = re.globalMatch(text);

    while (it.hasNext()) {
        QRegularExpressionMatch match = it.next();
        QString pty = match.captured(1);
        if (m_ptyA.isEmpty()) {
            m_ptyA = pty;
        } else if (m_ptyB.isEmpty() && pty != m_ptyA) {
            m_ptyB = pty;
        }
    }

    if (!m_ptyA.isEmpty() && !m_ptyB.isEmpty()) {
        m_waitingForPorts = false;
        qWarning() << "[MOUSART] Virtual port: A=" << m_ptyA << "(self) B=" << m_ptyB << "(external)";
        openSideA();
    }
}

void VirtualSerialManager::openSideA()
{
    m_fd = ::open(m_ptyA.toLocal8Bit().constData(), O_RDWR | O_NOCTTY | O_NONBLOCK);
    if (m_fd < 0) {
        emit errorOccurred(tr("无法打开虚拟串口A端: ") + QString::fromLocal8Bit(strerror(errno)));
        return;
    }

    struct termios tio;
    if (tcgetattr(m_fd, &tio) == 0) {
        cfmakeraw(&tio);
        tio.c_cc[VMIN] = 0;
        tio.c_cc[VTIME] = 1;
        tio.c_lflag &= ~(ECHO | ECHONL | ECHOE | ICANON);
        tio.c_oflag &= ~(ONLCR | OCRNL);
        tcsetattr(m_fd, TCSANOW, &tio);
    }

    int flags = fcntl(m_fd, F_GETFL, 0);
    if (flags >= 0) fcntl(m_fd, F_SETFL, flags & ~O_NONBLOCK);

    m_readNotifier = new QSocketNotifier(m_fd, QSocketNotifier::Read, this);
    connect(m_readNotifier, &QSocketNotifier::activated, this, &VirtualSerialManager::onFdReadyRead);
    m_readNotifier->setEnabled(true);

    resetStats();
    qWarning() << "[MOUSART] Side A opened:" << m_ptyA;
    emit isActiveChanged();
    emit externalPortChanged();
}

void VirtualSerialManager::onSocatFinished(int exitCode, QProcess::ExitStatus status)
{
    Q_UNUSED(exitCode)
    Q_UNUSED(status)

    if (m_readNotifier) {
        m_readNotifier->setEnabled(false);
        delete m_readNotifier;
        m_readNotifier = nullptr;
    }
    if (m_fd >= 0) {
        ::close(m_fd);
        m_fd = -1;
    }
    m_ptyA.clear();
    m_ptyB.clear();
    m_waitingForPorts = false;
    QFile::remove(VPORT_SYMLINK);
    emit isActiveChanged();
    emit externalPortChanged();
}

void VirtualSerialManager::onFdReadyRead(int fd)
{
    char buf[4096];
    ssize_t n = ::read(fd, buf, sizeof(buf));
    if (n > 0) {
        QByteArray data(buf, n);
        m_rxBytes += n;
        emit statsChanged();

        QString text = decodeData(data);
        emit dataReceived(text, data);
    }
}
#endif // !Q_OS_WIN

File: mousart/core/virtual_serial_manager.py
"""Virtual serial port management via socat (Linux only)."""
import os
import sys
import subprocess
import threading
import time
from mousart.qt_compat import *

from mousart.utils.encoding import decode_data


class VirtualSerialManager(QObject):
    """Manages a virtual serial port pair using socat on Linux."""

    is_active_changed = pyqtSignal()
    external_port_changed = pyqtSignal()
    data_received = pyqtSignal(str, bytes)
    error_occurred = pyqtSignal(str)
    timed_send_active_changed = pyqtSignal()
    timed_send_completed = pyqtSignal(str)
    stats_changed = pyqtSignal()
    echo_enabled_changed = pyqtSignal()
    newline_changed = pyqtSignal()
    receive_encoding_changed = pyqtSignal()

    VPORT_SYMLINK = "/tmp/mousart_vport"

    def __init__(self, parent=None):
        super().__init__(parent)
        self._socat_proc = None
        self._pty_a = ""
        self._pty_b = ""
        self._fd = -1
        self._active = False
        self._running = False
        self._read_thread = None

        # Timed send
        self._timed_send_timer = QTimer(self)
        self._timed_send_timer.timeout.connect(self._on_timed_send_tick)
        self._timed_send_data = ""
        self._timed_send_hex = False
        self._timed_send_count = -1
        self._timed_send_sent = 0

        # Stats
        self._rx_bytes = 0
        self._tx_bytes = 0
        self._rx_bytes_snapshot = 0
        self._tx_bytes_snapshot = 0
        self._stats_timer = QElapsedTimer()
        self._stats_timer.start()
        self._stats_update_timer = QTimer(self)
        self._stats_update_timer.setInterval(1000)
        self._stats_update_timer.timeout.connect(self._on_stats_timer)
        self._stats_update_timer.start()

        # Settings
        self._echo_enabled = False
        self._newline_cr = True
        self._newline_lf = True
        self._receive_encoding = "UTF-8"

    @pyqtProperty(bool, notify=is_active_changed)
    def isActive(self):
        return self._active

    @pyqtProperty(str, notify=external_port_changed)
    def externalPort(self):
        return self.VPORT_SYMLINK if self._active else ""

    # Stats
    @pyqtProperty(int, notify=stats_changed)
    def rxBytes(self):
        return self._rx_bytes

    @pyqtProperty(int, notify=stats_changed)
    def txBytes(self):
        return self._tx_bytes

    @pyqtProperty(float, notify=stats_changed)
    def rxRate(self):
        elapsed = self._stats_timer.elapsed()
        if elapsed < 100:
            return 0.0
        return (self._rx_bytes - self._rx_bytes_snapshot) * 1000.0 / elapsed

    @pyqtProperty(float, notify=stats_changed)
    def txRate(self):
        elapsed = self._stats_timer.elapsed()
        if elapsed < 100:
            return 0.0
        return (self._tx_bytes - self._tx_bytes_snapshot) * 1000.0 / elapsed

    @pyqtProperty(bool, notify=timed_send_active_changed)
    def timedSendActive(self):
        return self._timed_send_timer.isActive()

    # Settings
    @pyqtProperty(bool, notify=echo_enabled_changed)
    def echoEnabled(self):
        return self._echo_enabled

    @echoEnabled.setter
    def echoEnabled(self, v):
        if self._echo_enabled == v:
            return
        self._echo_enabled = v
        self.echo_enabled_changed.emit()

    @pyqtProperty(bool, notify=newline_changed)
    def newlineCr(self):
        return self._newline_cr

    @newlineCr.setter
    def newlineCr(self, v):
        if self._newline_cr == v:
            return
        self._newline_cr = v
        self.newline_changed.emit()

    @pyqtProperty(bool, notify=newline_changed)
    def newlineLf(self):
        return self._newline_lf

    @newlineLf.setter
    def newlineLf(self, v):
        if self._newline_lf == v:
            return
        self._newline_lf = v
        self.newline_changed.emit()

    @pyqtProperty(str, notify=receive_encoding_changed)
    def receiveEncoding(self):
        return self._receive_encoding

    @receiveEncoding.setter
    def receiveEncoding(self, v):
        if self._receive_encoding == v:
            return
        self._receive_encoding = v
        self.receive_encoding_changed.emit()

    @pyqtSlot(result=bool)
    def startVirtualPort(self):
        """Start a virtual serial port pair using socat."""
        if sys.platform == "win32":
            self.error_occurred.emit("虚拟串口功能在 Windows 上暂不可用，请使用硬件串口调试模式。")
            return False

        if self._active:
            self.stopVirtualPort()

        try:
            # Remove old symlink
            if os.path.islink(self.VPORT_SYMLINK):
                os.unlink(self.VPORT_SYMLINK)

            self._socat_proc = subprocess.Popen(
                ["socat", "-d", "-d", "pty,raw,echo=0",
                 f"pty,raw,echo=0,link={self.VPORT_SYMLINK}"],
                stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                text=True
            )

            # Wait for socat to create PTYs
            import re
            pty_paths = []
            start_time = time.time()
            while time.time() - start_time < 3.0:
                # Read stderr for PTY info
                import select
                ready, _, _ = select.select([self._socat_proc.stderr], [], [], 0.1)
                if ready:
                    line = self._socat_proc.stderr.readline()
                    if line:
                        match = re.search(r"PTY is (\S+)", line)
                        if match:
                            pty_paths.append(match.group(1))
                            if len(pty_paths) >= 2:
                                break
                if self._socat_proc.poll() is not None:
                    break

            if len(pty_paths) < 2:
                self.stopVirtualPort()
                self.error_occurred.emit("启动 socat 失败: 无法获取 PTY 路径")
                return False

            self._pty_a = pty_paths[0]
            self._pty_b = pty_paths[1]

            # Open side A
            import termios
            self._fd = os.open(self._pty_a, os.O_RDWR | os.O_NOCTTY)

            # Configure terminal
            attrs = termios.tcgetattr(self._fd)
            import tty
            tty.setraw(self._fd)
            attrs = termios.tcgetattr(self._fd)
            attrs[6][termios.VMIN] = 0
            attrs[6][termios.VTIME] = 1
            attrs[3] &= ~(termios.ECHO | termios.ECHONL | termios.ECHOE | termios.ICANON)
            attrs[1] &= ~(termios.ONLCR | termios.OCRNL)
            termios.tcsetattr(self._fd, termios.TCSANOW, attrs)

            self._active = True
            self._running = True
            self._read_thread = threading.Thread(target=self._read_loop, daemon=True)
            self._read_thread.start()

            self.resetStats()
            self.is_active_changed.emit()
            self.external_port_changed.emit()
            return True

        except FileNotFoundError:
            self.error_occurred.emit("socat 未安装。请运行: sudo apt install socat")
            return False
        except Exception as e:
            self.error_occurred.emit(f"启动虚拟串口失败: {e}")
            self.stopVirtualPort()
            return False

    @pyqtSlot()
    def stopVirtualPort(self):
        """Stop the virtual serial port."""
        self.stopTimedSend()
        self._running = False

        if self._fd >= 0:
            try:
                os.close(self._fd)
            except OSError:
                pass
            self._fd = -1

        if self._socat_proc:
            try:
                self._socat_proc.terminate()
                self._socat_proc.wait(timeout=2)
            except Exception:
                try:
                    self._socat_proc.kill()
                except Exception:
                    pass
            self._socat_proc = None

        self._pty_a = ""
        self._pty_b = ""

        if os.path.islink(self.VPORT_SYMLINK):
            try:
                os.unlink(self.VPORT_SYMLINK)
            except OSError:
                pass

        was_active = self._active
        self._active = False
        if was_active:
            self.is_active_changed.emit()
            self.external_port_changed.emit()

    @pyqtSlot(str, bool, result=int)
    def sendData(self, data: str, hex_mode: bool) -> int:
        if self._fd < 0:
            self.error_occurred.emit("虚拟串口未打开")
            return -1

        try:
            if hex_mode:
                hex_str = data.replace(" ", "").replace("\n", "").replace("\r", "")
                data_bytes = bytes.fromhex(hex_str)
            else:
                text = data
                if self._newline_cr:
                    text += "\r"
                if self._newline_lf:
                    text += "\n"
                data_bytes = text.encode("utf-8")

            written = os.write(self._fd, data_bytes)
            self._tx_bytes += written
            self.stats_changed.emit()
            return written
        except Exception as e:
            self.error_occurred.emit(str(e))
            return -1

    @pyqtSlot(str, bool, int, int)
    def startTimedSend(self, data: str, hex_mode: bool, interval_ms: int, count: int = -1):
        if not self._active:
            self.error_occurred.emit("虚拟串口未打开")
            return
        self._timed_send_data = data
        self._timed_send_hex = hex_mode
        self._timed_send_count = count
        self._timed_send_sent = 0
        self._timed_send_timer.setInterval(max(interval_ms, 1))
        self._timed_send_timer.start()
        self.timed_send_active_changed.emit()

    @pyqtSlot()
    def stopTimedSend(self):
        self._timed_send_timer.stop()
        self._timed_send_sent = 0
        self._timed_send_count = -1
        self.timed_send_active_changed.emit()

    @pyqtSlot()
    def resetStats(self):
        self._rx_bytes = 0
        self._tx_bytes = 0
        self._rx_bytes_snapshot = 0
        self._tx_bytes_snapshot = 0
        self._stats_timer.restart()
        self.stats_changed.emit()

    def _read_loop(self):
        """Background thread for reading from virtual serial port."""
        import select
        while self._running and self._fd >= 0:
            try:
                ready, _, _ = select.select([self._fd], [], [], 0.05)
                if ready:
                    data = os.read(self._fd, 4096)
                    if data:
                        self._rx_bytes += len(data)
                        self.stats_changed.emit()
                        text = decode_data(data, self._receive_encoding)
                        self.data_received.emit(text, data)
                    elif not data:
                        # EOF
                        break
            except OSError:
                break
            except Exception:
                break

    def _on_timed_send_tick(self):
        if not self._active:
            self.stopTimedSend()
            return
        result = self.sendData(self._timed_send_data, self._timed_send_hex)
        if result > 0:
            self._timed_send_sent += 1
            self.timed_send_completed.emit(self._timed_send_data)
            if 0 < self._timed_send_count <= self._timed_send_sent:
                self.stopTimedSend()

    def _on_stats_timer(self):
        self._rx_bytes_snapshot = self._rx_bytes
        self._tx_bytes_snapshot = self._tx_bytes
        self._stats_timer.restart()
        self.stats_changed.emit()
